internal/server/allocate_cases.go

```diff
@@ -0,0 +1,134 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/ministryofjustice/opg-go-common/template"
+	"github.com/ministryofjustice/opg-sirius-lpa-frontend/internal/sirius"
+	"golang.org/x/sync/errgroup"
+)
+
+type AllocateCasesClient interface {
+	AllocateCases(ctx sirius.Context, assigneeID int, allocations []sirius.CaseAllocation) error
+	Case(ctx sirius.Context, id int) (sirius.Case, error)
+	Teams(ctx sirius.Context) ([]sirius.Team, error)
+}
+
+type allocateCasesData struct {
+	XSRFToken string
+	Entities  []string
+	Success   bool
+	Error     sirius.ValidationError
+
+	Teams            []sirius.Team
+	AssignTo         string
+	AssigneeUserName string
+}
+
+func AllocateCases(client AllocateCasesClient, tmpl template.Template) Handler {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		if err := r.ParseForm(); err != nil {
+			return err
+		}
+
+		var caseIDs []int
+		for _, id := range r.Form["id"] {
+			caseID, err := strconv.Atoi(id)
+			if err != nil {
+				return err
+			}
+			caseIDs = append(caseIDs, caseID)
+		}
+
+		if len(caseIDs) == 0 {
+			return errors.New("id missing")
+		}
+
+		ctx := getContext(r)
+		data := allocateCasesData{XSRFToken: ctx.XSRFToken}
+
+		group, groupCtx := errgroup.WithContext(ctx.Context)
+
+		group.Go(func() error {
+			teams, err := client.Teams(ctx.With(groupCtx))
+			if err != nil {
+				return err
+			}
+
+			data.Teams = teams
+			return nil
+		})
+
+		var casesMu sync.Mutex
+		var allocations []sirius.CaseAllocation
+
+		for _, caseID := range caseIDs {
+			caseID := caseID
+
+			group.Go(func() error {
+				caseitem, err := client.Case(ctx.With(groupCtx), caseID)
```

Comment:
It seems a waste having to do lookups for every case just to get the case type. Maybe we could (in the future) refactor the allocation API endpoint to not require them and make this unnecessary?

Reply:
The only reason I thought this was OK is that we're getting the cases to show the UID on the page of the one's selected. Although one thing I could do is not requery them for the POST call, and pass all the info in from the form